collect_app/src/main/java/org/odk/collect/android/formentry/backgroundlocation/FormEntryViewModel.java

```diff
@@ -0,0 +1,49 @@
+package org.odk.collect.android.formentry.backgroundlocation;
+
+import androidx.annotation.NonNull;
+import androidx.lifecycle.ViewModel;
+
+import org.odk.collect.android.application.Collect;
+import org.odk.collect.android.location.client.GoogleLocationClient;
+
+public class FormEntryViewModel extends ViewModel {
+    @NonNull
+    private final BackgroundLocationManager locationManager;
+
+    public FormEntryViewModel() {
```

Comment:
How'd you feel about the `BackgroundLocationManager` being a constructor level dependency here? I think it would make it easier to pull apart/test in future.

Reply:
Thanks for bringing that up, I should have called it out explicitly. One of my design goals was to make the view model layer so thin that it wouldn't need its own testing. I see its role as maintaining model objects through the activity lifecycle and passing main thread actions to model objects. In a case where there might be live-updating data (form lists!), it could act as a listener for those model objects and present a livedata object to the activity. 

That way, we don't have to worry about constructor dependencies, factories, creating the dependencies at the right time so they don't get recreated on orientation changes, etc. Does that make sense? Or is there a way to manage constructor-level dependencies that is simpler?

Reply:
So if we use a `ViewModel` factory to construct the view model we will actually get the same view model instance between orientation changes so we should be ok.

I think this is something I might just be overly nitpicky about but I generally like having anything your object depends on come in at a constructor level to make it easier to comprehend what things that object is going to interact with. It's also a refactor you're pretty much always going to have to carry out if you get the object to a point that you feel you'd like unit tests for it.

Reply:
I agree with you that in general constructor-level dependencies are better. But I can't figure it out in this case! What class would you give the responsibility of building the dependencies to? The best I can come up with is this in `onCreate` but the dependencies are recreated on orientation changes which I definitely don't want:

```java
       ViewModelProvider.Factory factory = new ViewModelProvider.Factory() {
            @Override
            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
                if (modelClass.equals(FormEntryViewModel.class)) {
                    GoogleLocationClient googleLocationClient = new GoogleLocationClient(Collect.getInstance().getApplicationContext());

                    BackgroundLocationManager locationManager =
                            new BackgroundLocationManager(googleLocationClient, new BackgroundLocationHelper());
                    return (T) new FormEntryViewModel(locationManager);
                }
                return null;
            }
        };

        viewModel = ViewModelProviders.of(this, factory).get(FormEntryViewModel.class);
```

Same issue if the activity has that responsibility and passes them into the factory.

The only thing I can think of is to make them (or the factory) Application-level singletons which I'm really not excited about. Somehow I can't find any helpful examples or tutorials!

Giving that responsibility to the viewmodel gives the dependencies exactly the lifecycle I want. As long as the viewmodel stays really thin with no logic that needs testing, that seems like a reasonable compromise to me.

Reply:
I agree that constructor-level dependencies are best. Is 2e61ad2 what you had in mind? Now the activity builds its own factory dependency. I think that's ok because the activity is responsible for building many of its own dependencies already anyway.

If you saw a previous confused comment, for a minute it looked like `create` was being called multiple times and I couldn't figure out the point of having a factory. 🥴